pkg/rpc/rpc.go

```diff
@@ -112,48 +115,39 @@ func (r *RPC) Call(method string, params json.RawMessage) (json.RawMessage, erro
 func (r *RPC) Err() error {
 	select {
 	case <-r.Done():
-		return r.err
+		return r.processResponsesErr
 	default:
 		return nil
 	}
 }
 
-// Done sends when the errChannel has been closed.
-// errChannel is closed when encoder/decoder fails to write/read
+// Done sends when the processResponsesErrCh has been closed.
+// processResponsesErrCh is closed when decoder fails to read
 func (r *RPC) Done() <-chan struct{} {
-	return r.errChannel
+	return r.processResponsesErrCh
 }
 
 // Close closes the RPC.writer.
-// Further requests leads encoder/decoder to fail
+// All further calls to Call lead to an error.
+// The Process will be terminated as soon as all pending requests have been processed.
 func (r *RPC) Close() error {
 	r.encoderMu.Lock()
 	defer r.encoderMu.Unlock()
 
-	return r.writer.Close()
-}
-
-// setErr sets the err encoder/decoder fails and closes errChannel
-func (r *RPC) setErr(err error) {
-	r.errOnce.Do(func() {
-		if reqCount := len(r.pendingRequests); reqCount > 0 {
-			r.logger.Debugf("Cancelling %d pending request(s)", reqCount)
-		}
+	r.encoder = nil
 
-		r.err = &Error{cause: err}
-		close(r.errChannel)
-	})
+	return r.writer.Close()
 }
 
 // processResponses sends responses to its channel (identified by response.id)
 // In case of any error, all pending requests are dropped
 func (r *RPC) processResponses() {
 	for r.Err() == nil {
 		var response Response
-		err := r.decoder.Decode(&response)
-
-		if err != nil {
-			r.setErr(fmt.Errorf("failed to read json response: %w", err))
+		if err := r.decoder.Decode(&response); err != nil {
+			r.processResponsesErr = &Error{cause: fmt.Errorf("failed to read json response: %w", err)}
+			_ = r.Close()
+			close(r.processResponsesErrCh)
```

Comment:
I'd swap those two operations around. Closing the channel can always proceed immediately whereas `r.Close()` might have to wait for the lock for a short while.